EIPS/eip-2386.md

```diff
@@ -0,0 +1,210 @@
+discussions-to: https://ethereum-magicians.org/t/eip-2386-walletstore/3792
+status: Draft
+type: Standards Track
+category: ERC
+created: 2019-11-21
+requires: 2333, 2334
+---
+
+## Simple Summary
+
+A JSON format for the storage and retrieval of Ethereum 2 wallet definitions.
+
+## Abstract
+
+Ethereum has the concept of keystores: pieces of data that define a key (see [EIP-2333](https://eips.ethereum.org/EIPS/eip-2333) for details).  This adds the concept of walletstores: stores that define wallets and how keys in said wallets can be created.
+
+## Motivation
+
+There are multiple different methods to generate private keys.  They may, for example, be generated from random data (_i.e._ non-deterministically) or from seed data and a deterministic path (_i.e._ hierarchical deterministic).  Historically, individual keys have been generated without reference to each other, however the idea of the wallet being both a grouping of keys and a definition of how future keys will be generated allows a higher degree of flexibility.  A single user could have multiple wallets that have different levels of security on them, different key generation methods, _etc._  If a user has multiple wallets they need identification, both programmatically and in a user-friendly fashion.
+
+Given that a wallet has an amount of data and metadata that is useful when accessing existing keys and creating new keys, standardizing this information and how it is stored allows it to be portable between different wallet providers with minimal effort.
+
+## Specification
+
+The elements of a walletstore are as follows:
+
+### UUID
+
+The `uuid` provided in the walletstore is a randomly-generated type 4 UUID as specified by [RFC 4122](https://tools.ietf.org/html/rfc4122). It is intended to be used as a 128-bit proxy for referring to a particular wallet, used to uniquely identify wallets.
+
+This element MUST be present.  It MUST be a string following the syntactic structure as laid out in [section 3 of RFC 4122](https://tools.ietf.org/html/rfc4122#section-3).
+
+### Name
+
+The `name` provided in the walletstore is a UTF-8 string.  It is intended to serve as the user-friendly accessor.  The only restriction on the name is that it MUST NOT start with the underscore (`_`) character.
+
+This element MUST be present.  It MUST be a string.
+
+### Version
+
+The `version` provided is the version of the walletstore.
+
+This element MUST be present.  It MUST be the integer `1`.
+
+### Type
+
+The `type` provided is the type of wallet.  This informs mechanisms such as key generation.
+
+This element MUST be present.  It MUST be one of the following two strings:
+
+  - `non-deterministic`: defined for wallets that create private keys from random information
+  - `hierarchical deterministic`: defined for wallets that create private keys based on the hierarchical deterministic method as outlined in [EIP-2334](https://eips.ethereum.org/EIPS/eip-2334)
+
+### Crypto
```

Comment:
I've noticed that there's perhaps some wiggle room for interpretation here.

This is how I've assumed the flow works here:

- Generate some